### PR TITLE
check sampler type in as_bind_group derives

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

currently if we use an image with the wrong sampler type in a material, wgpu panics with an invalid texture format. turn this into a warning and fail more gracefully.

## Solution

the expected sampler type is specified in the AsBindGroup derive, so we can just check the image sampler is what it should be.

i am not totally sure about the mapping of image sampler type to #[sampler(type)], i assumed:

```
    "filtering" => [ TextureSampleType::Float { filterable: true } ],
    "non_filtering" => [
        TextureSampleType::Float { filterable: false },
        TextureSampleType::Sint,
        TextureSampleType::Uint,
    ],
    "comparison" => [ TextureSampleType::Depth ],
```

### CODE REVIEW

File: crates/bevy_render/src/render_resource/bind_group.rs
Comment (afonsolage) [inline crates/bevy_render/src/render_resource/bind_group.rs:357]: I think it would be better to avoid too broad terms, since `InvalidData` can be anything. In this case, the new error is added specifically for when missing or invalid sampler, so it could be something in the likes of `InvalidSampler`

**Review by afonsolage — COMMENTED** [review]
I've run into this issue and I agree this shouldn't be a hard error, since it's totally recoverable, by just changing the sample or data format.

**Review by kristoff3r — APPROVED** [review]

**Review by IceSentry — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@robtfm I'll merge this once the merge conflicts are resolved :)

**Comment by robtfm** [comment]
apologies, i destroyed the history trying to rebase...

**Comment by alice-i-cecile** [comment]
That's git for you ;) No worries, it gets squashed to the PR level anyways.

**Pull request merged**
